proposals/text/0000-shaper-in-prompt-template.md

```diff
@@ -0,0 +1,153 @@
+            "If the documents do not contain the answer to the question, say that ‘answering is not possible given the available information.’\n"
+            "$documents \n Question: $query; Answer: ",
+            input_shapers=[
+                Shaper(
+                    func="join_documents",
+                    inputs={
+                        "documents": "documents",
+                        "delimiter": "delimiter",
+                        "pattern": "pattern",
+                        "str_replace": "str_replace",
+                    },
+                    outputs=["documents"],
+                    params={
+                        "delimiter": "\n",
+                        "pattern": "\nDocument[$idx]: $content",
+                        "str_replace": {"\n": " ", "[": "(", "]": ")"},
+                    },
+                ),
+                Shaper(func="value_to_list", inputs={"value": "query", "target_list": "documents"}, outputs=["query"]),
+            ],
+            output_shapers=[
+                Shaper(
+                    func="strings_to_answers",
+                    inputs={"strings": "results", "documents": "documents"},
+                    outputs=["answers"],
+                )
+            ],
+            output_variable="answers",
+        )
+    ```
+
+As it's crucial to iterate quickly on prompts, shaper params can be overridden by PromptNode like this:
+
+    ```python
+    pn = PromptNode(default_prompt_template="question-answering-with-references", shaper_params={"delimiter": "MY_NEW_DELIMITER"})
+    ```
+
+This would set the default delimiter `\n` as defined in the template to `MY_NEW_DELIMITER` for this specific PromptNode.
+
+# Motivation
+
+Currently using PromptNode is a bit cumbersome as:
+- for using it in popular use-cases like question-answering, it requires to add the Shapers to the pipeline manually which creates a lot of boilerplate code and is not very intuitive
+- to customize a prompt within a pipeline, you may need to change four different things: the prompt node, the prompt template, the input shapers and the output shapers. This is not ideal as it requires to write a lot of boilerplate code and makes it hard to iterate quickly on prompts.
+- if you wanted to share your prompt template with the community, you would need to share the whole pipeline (as you do need shapers), which is not ideal as it may contain other nodes that are not relevant to the prompt template.
+
+
+# Detailed design
+
+PromptTemplate gets two new attributes: `input_shapers` and `output_shapers`. These are lists of Shaper objects that are applied to the input and output of the prompt respectively. The `input_shapers` are applied to the input of the prompt, the `output_shapers` are applied to the output of the prompt:
+PromptNode calls `PromptTemplate.prepare` before exectuting the prompt. `PromptTemplate.prepare` makes all `input_shapers` run on the `invocation_context`.
```

Comment:
Let's just call it pre_process for naming consistency?

Reply:
I think preprocess would bring in mind a bit different concept imho, but I don't have strong opinions here